src/validate-icon.c

```diff
@@ -36,37 +38,67 @@
 #define ICON_VALIDATOR_GROUP "Icon Validator"
 #define MAX_ICON_SIZE 512
 #define MAX_SVG_ICON_SIZE 4096
+#define BUFFER_SIZE 4096
 
 static int
-validate_icon (const char *filename)
+validate_icon (int input_fd)
 {
-  GdkPixbufFormat *format;
-  int max_size;
-  int width, height;
-  g_autofree char *name = NULL;
   const char *allowed_formats[] = { "png", "jpeg", "svg", NULL };
-  g_autoptr(GdkPixbuf) pixbuf = NULL;
+  g_autoptr(GBytes) bytes = NULL;
   g_autoptr(GError) error = NULL;
+  GdkPixbufFormat *format;
   g_autoptr(GKeyFile) key_file = NULL;
   g_autofree char *key_file_data = NULL;
+  g_autoptr(GdkPixbufLoader) loader = NULL;
+  g_autoptr(GMappedFile) mapped = NULL;
+  int max_size, width, height;
+  g_autofree char *name = NULL;
+  GdkPixbuf *pixbuf;
 
-  format = gdk_pixbuf_get_file_info (filename, &width, &height);
-  if (format == NULL)
+  if (!(mapped = g_mapped_file_new_from_fd (input_fd, FALSE, &error)))
```

Comment:
Wondering if this also should use `XdpSealedFd`, but I guess there isn't really any benefit and would make it harder to call the program from a shell.

Reply:
i don't think that's important. If we want to improve the validator even more i would to it later and switch to https://gitlab.gnome.org/sophie-h/glycin